### PR TITLE
[Feature/Identity] Introduce Identity module

### PR DESCRIPTION
**Pull request by cwperks**
### Description

Opening a PR to solicit feedback for introduction of Identity module which includes an implementation of Basic Auth.

This PR introduces a new sandbox module **identity** that will use some of the existing extension points that the security plug-in does to authenticate rest requests and pass a token around on the header of the threadcontext of a task that identifies the user and can subsequently be used for authorization.

This new identity module uses a few existing extension points from the `ActionPlugin` and the `NetworkPlugin`.

From the `ActionPlugin` this branch uses:

- `getRestHandlerWrapper` to provide a wrapper that handles authentication. As of now, there is only a Basic auth mechanism that uses the internal IdP in this feature branch to authenticate the user and return a 403 if the request cannot be authenticated
- `getActionFilters` - This branch introduces an AuthorizationFilter that is intended to be used to perform authorization. This is mostly pass-through at the moment and right now it verifies that a token is present and valid before the TransportRequest performs its `doExecute`

From the `NetworkPlugin` this uses:

- `getTransportInterceptors` - The transport interceptor intercepts outgoing TransportRequests and can modify the request before its sent to another node. When testing this branch, I ran into problems with how the `TransportMessageListener` intercepted outgoing requests as the ThreadContext was not available to inspect to ensure that the node that received the RestRequest created a token before sending the transport request to other nodes. When running the test its clear to see that other nodes received the created token, but the message listener is unable to get it because of how its wrapped in an `ActionListener` in `OutboundHandler`:

```
void sendRequest(...) throws IOException, TransportException {
        Version version = Version.min(this.version, channelVersion);
        OutboundMessage.Request message = new OutboundMessage.Request(
            threadPool.getThreadContext(),
            features,
            request,
            version,
            action,
            requestId,
            isHandshake,
            compressRequest
        );
        ActionListener<Void> listener = ActionListener.wrap(() -> messageListener.onRequestSent(node, requestId, action, request, options));
        sendMessage(channel, message, listener);
    }
```

The transport interceptor has access to the ThreadContext and the tests will be updated to use the interceptor.

### Check List
- [X] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: sandbox/libs/authn/src/main/resources/shiro.ini
Comment (DarshitChanpura) [inline sandbox/libs/authn/src/main/resources/shiro.ini:29]: Do we need authcBasic here?

File: sandbox/modules/identity/src/internalClusterTest/java/org/opensearch/identity/BasicAuthenticationIT.java
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/internalClusterTest/java/org/opensearch/identity/BasicAuthenticationIT.java:150]: if not needed, we can remove commented code

File: sandbox/modules/identity/src/internalClusterTest/java/org/opensearch/identity/HttpSmokeTestCaseWithIdentity.java
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/internalClusterTest/java/org/opensearch/identity/HttpSmokeTestCaseWithIdentity.java:26]: for my understanding, we should add a description of why we need this class instead of using OpenSearchIntegTestCase directly for BasicAuthenticationIT.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/PrivilegesEvaluatorResponse.java
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/PrivilegesEvaluatorResponse.java:17]: Can you add a brief javadoc for this and other newly introduced classes in this PR?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityFilter.java
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityFilter.java:105]: this check would always be true based on setting on line 99. Is that expected or should line 99 be changed to something conditional?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityRestFilter.java
Comment (DarshitChanpura) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityRestFilter.java:91]: Would logging all these info per request have an effect on performance? I'm not sure

File: server/src/main/java/org/opensearch/rest/RestController.java
Comment (DarshitChanpura) [inline server/src/main/java/org/opensearch/rest/RestController.java:90]: would this variable be used in this class?

**Review by DarshitChanpura — COMMENTED** [review]
Thank you for moving identity project to a module @cwperks !

File: plugins/discovery-azure-classic/licenses/commons-lang-2.6.jar.sha1
Comment (cwperks) [reply plugins/discovery-azure-classic/licenses/commons-lang-2.6.jar.sha1:1]: Yes, I removed the `server` dependency on the `:sandbox:libs:opensearch-authn` library and ran `updateSHAs` after removing the dependency.

File: sandbox/libs/authn/build.gradle
Comment (cwperks) [reply sandbox/libs/authn/build.gradle:27]: I needed to add these to get the `IniSecurityManagerFactory` working.

File: sandbox/libs/authn/src/main/java/org/opensearch/authn/internal/InternalSubject.java
Comment (cwperks) [reply sandbox/libs/authn/src/main/java/org/opensearch/authn/internal/InternalSubject.java:70]: This section needs to be revisited for performance to see if the entire `./gradlew check` can run with identity enabled. I can remove the commented out line for now.

File: sandbox/libs/authn/docs/auth_flow.md
Comment (cwperks) [reply sandbox/libs/authn/docs/auth_flow.md]: Added

File: sandbox/modules/identity/src/internalClusterTest/java/org/opensearch/identity/HttpSmokeTestCaseWithIdentity.java
Comment (cwperks) [reply sandbox/modules/identity/src/internalClusterTest/java/org/opensearch/identity/HttpSmokeTestCaseWithIdentity.java:26]: I was running into issues where the `SecurityRestFilter` (used in the `getRestHandlerWrapper` extension point) was not getting invoked when running requests with `OpenSearchIntegTestCase`. This class overrides `addMockHttpTransport` to set it to false, adds IdentityPlugin and enables the module.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/PrivilegesEvaluatorResponse.java
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/PrivilegesEvaluatorResponse.java:17]: Sure, some of these classes are boilerplate we can use when visiting authorization in future PRs. The `SecurityFilter` in this PR can be used for authorization, but in this case I added these when testing that tokens were being passed internally with transport requests by logging them out in the class. The test to verify that tokens are transported internally uses a `TransportMessageListener` to intercept messages when they are received at different nodes in the cluster.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityFilter.java
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityFilter.java:105]: That is more boilerplate that is meant to be expanded on when authorization is tackled. When I started working on this branch one of the big ideas was to deprecate the usage of serializing the User object and transporting around the cluster on the thread context headers as a proxy for a request that has previously been authenticated on the rest layer. Instead of serializing the user object, this PR introduces tokens for authentication information and transports them around the cluster using the thread context headers. I originally wanted to terminate an internal transport request if it did not contain authentication information with it. The problem I faced is that this header would be populated for internal transport requests that originated from the Rest Layer, but it was not present for internal actions invoked from elsewhere.

In the future PR we need to investigate where internal actions are invoked from and choose whether we should send information indicating that its starting from an internal source.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityRestFilter.java
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityRestFilter.java:91]: I can switch these log statements to debug, this was useful to see what was happening in the cluster but needs to be addressed before merge into main.

File: server/src/main/java/org/opensearch/rest/RestController.java
Comment (cwperks) [reply server/src/main/java/org/opensearch/rest/RestController.java:90]: No, it should be moved.

File: sandbox/libs/authn/src/main/resources/shiro.ini
Comment (cwperks) [reply sandbox/libs/authn/src/main/resources/shiro.ini:29]: Not currently, but this is an area we should look into further: https://shiro.apache.org/web.html

Removing this.

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityRestFilter.java
Comment (DarshitChanpura) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityRestFilter.java]: Can you add a Todo here so we can track it?

**Review by DarshitChanpura — APPROVED** [review]
Thank you @cwperks !

**Comment by DarshitChanpura** [comment]
@saratvemulapalli Can you please help merge this?

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityRestFilter.java
Comment (cwperks) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/SecurityRestFilter.java]: Changed to debug. Run the tests with `-Dtests.opensearch.logger.level=DEBUG` to see the output. This output helped with debugging the tests I wrote for this PR.

File: server/src/main/java/org/opensearch/rest/RestController.java
Comment (cwperks) [reply server/src/main/java/org/opensearch/rest/RestController.java]: Removed

**Review by saratvemulapalli — APPROVED** [review]
Didnt want to block you work. 
But i'd like to review the change sometime this week. 
@cwperks I'll merge it and i'll drop comments if I have any.

**Pull request merged**
